### ISSUE TITLE
.NET Core 2.1 application fails to start on Linux (Ubuntu) cluster.

### ISSUE THREAD

**Issue by git-heath** [Azure/service-fabric-issues#1178]
I have a simple rest service deployed to an ubuntu service fabric cluster which starts perfectly when using .NET core version 2.0 but fails with the following error with version 2.1. Is 2.1 officially supported for a linux cluster ? I could not find this information anywhere. Note that 2.1 works on our Windows cluster.

`Unhealthy event: SourceId='System.RA', Property='ReplicaOpenStatus', HealthState='Warning', ConsiderWarningAsError=false. Replica had multiple failures during open on _NodeType2_2. API call: IStatelessServiceInstance.Open(); Error = System.ArgumentException (-2147024809) Item has already been added. Key in dictionary: 'LD_LIBRARY_PATH' Key being added: 'LD_LIBRARY_PATH' at System.Collections.Hashtable.Insert(Object key, Object nvalue, Boolean add) at System.Environment.ToHashtable(IEnumerable1 pairs)
at Microsoft.Extensions.Configuration.EnvironmentVariables.EnvironmentVariablesConfigurationProvider.Load()
at Microsoft.Extensions.Configuration.ConfigurationRoot..ctor(IList1 providers) at Microsoft.Extensions.Configuration.ConfigurationBuilder.Build() at Microsoft.AspNetCore.Hosting.WebHostBuilder..ctor() at HelloWorld.HelloWorld.<>c__DisplayClass1_0.b__1(String url, AspNetCoreCommunicationListener listener) in c:\users\davies\source\repos\HelloWorldCloud\HelloWorld\HelloWorld.cs:line 38 at Microsoft.ServiceFabric.Services.Communication.AspNetCore.AspNetCoreCommunicationListener.OpenAsync(CancellationToken cancellationToken) at Microsoft.ServiceFabric.Services.Runtime.StatelessServiceInstanceAdapter.OpenCommunicationListenersAsync(CancellationToken cancellationToken) at Microsoft.ServiceFabric.Services.Runtime.StatelessServiceInstanceAdapter.System.Fabric.IStatelessServiceInstance.OpenAsync(IStatelessServicePartition partition, CancellationToken cancellationToken) For more information see: http://aka.ms/sfhealth`

**Comment by ashishnegi** [comment]
Exception is from within : `Microsoft.AspNetCore.Hosting.WebHostBuilder` i.e. `AspNetCore`.
so, your app is failing to start.

Is your SF app standalone .net core app or framework dependent FDD ?
If it's FDD, how are you making sure 2.1 dotnet runtime is present on cluster machines.

Are are you starting the app ? bash script ? Are you modifying the `LD_LIBRARY_PATH` somewhere in app code or your scripts ?

**Comment by git-heath** [comment]
The app is a standalone .NET core application. For starting the app, I am simply using Visual studio to do the deployment/startup of the application (Publish option on SF project) so no I have not modified anything to do with the target environment.

**Comment by ashishnegi** [comment]
@git-heath Please share your applicationmanifest.xml and servicemanifest.xml at asnegi[at]microsoft.com or attach here.

How are you starting the application in ServiceManifest.xml ? If its some bash script, also share it.

**Comment by git-heath** [comment]
@ashishnegi there is no bash script, the ExeHost simply points to the generated executable. Please find attached the two files.

`<?xml version="1.0" encoding="utf-8"?>
<ServiceManifest Name="HelloWorldPkg"
                 Version="1.0.0"
                 xmlns="http://schemas.microsoft.com/2011/01/fabric"
                 xmlns:xsd="http://www.w3.org/2001/XMLSchema"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <ServiceTypes>
    
    <StatelessServiceType ServiceTypeName="HelloWorldType" />
  </ServiceTypes>

  
  <CodePackage Name="Code" Version="1.0.3">
    <EntryPoint>
      <ExeHost>
        <Program>HelloWorld</Program>
        <WorkingFolder>CodePackage</WorkingFolder>
      </ExeHost>
    </EntryPoint>
  </CodePackage>

  
  <ConfigPackage Name="Config" Version="1.0.0" />

  <Resources>
    <Endpoints>
      
      <Endpoint Protocol="http" Name="ServiceEndpoint" Type="Input" Port="8875" />
    </Endpoints>
  </Resources>
</ServiceManifest>`

`<?xml version="1.0" encoding="utf-8"?>
<ApplicationManifest xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" ApplicationTypeName="HelloWorldCloudType" ApplicationTypeVersion="1.0.0" xmlns="http://schemas.microsoft.com/2011/01/fabric">
  <Parameters>
    <Parameter Name="HelloWorld_InstanceCount" DefaultValue="-1" />
  </Parameters>
  
  <ServiceManifestImport>
    <ServiceManifestRef ServiceManifestName="HelloWorldPkg" ServiceManifestVersion="1.0.0" />
    <ConfigOverrides />
  </ServiceManifestImport>
  <DefaultServices>
    
    <Service Name="HelloWorld" ServicePackageActivationMode="ExclusiveProcess">
      <StatelessService ServiceTypeName="HelloWorldType" InstanceCount="[HelloWorld_InstanceCount]">
        <SingletonPartition />
      </StatelessService>
    </Service>
  </DefaultServices>
</ApplicationManifest>`

**Comment by ashishnegi** [comment]
@git-heath The xmls look fine. You will need to debug this in your code. Try printing out env variables before `Microsoft.Extensions.Configuration.EnvironmentVariables.EnvironmentVariablesConfigurationProvider.` is called. And see who is modifying `LD_LIBRARY_PATH` and if there are multiple of these.

**Comment by git-heath** [comment]
@ashishnegi Ok I will try this. I also noticed that there was never a specific announcement from MS to state that anything greater than 2.0 is supported on Linux clusters. I found this announcement concerning 2.0.

https://azure.microsoft.com/en-us/roadmap/reliable-services-and-reliable-actors-on-linux-now-generally-available/

And also notice that the associated github "getting started" project for Linux only targets 2.0:

https://github.com/Azure-Samples/service-fabric-dotnet-core-getting-started

From your comments it appears that 2.1 should work but I feel there is a lack of documentation from MS to support this supposition.

**Comment by ashishnegi** [comment]
@git-heath since this is a self contained app, it is like any other executable from Service Fabric perspective. SF Platform does not know whether it is .net application or golang or rust or python program.

About SDK, in my understanding, 2.0 binaries are compatible with 2.1 runtime. So, you should not have any issue with respect to that, ideally.

Lets first figure out what is happening with `LD_LIBRARY_PATH` and asp .net and .net core 2.1.

**Comment by pksorensen** [comment]
@ashishnegi - i am experience the same thing.

i am using dotnet myapp.dll when it starts and its a framework dependent app. 

Further more, just doing Environment.GetEnvironmentVariables() also can make it throw.


```
System.ArgumentException: Item has already been added. Key in dictionary: 'LD_LIBRARY_PATH'  Key being added: 'LD_LIBRARY_PATH'
   at System.Collections.Hashtable.Insert(Object key, Object nvalue, Boolean add)
   at System.Environment.ToHashtable(IEnumerable`1 pairs)
   at Microsoft.Extensions.Configuration.EnvironmentVariables.EnvironmentVariablesConfigurationProvider.Load()
   at Microsoft.Extensions.Configuration.ConfigurationRoot..ctor(IList`1 providers)
   at Microsoft.Extensions.Configuration.ConfigurationBuilder.Build()
   at SInnovations.ServiceFabric.RegistrationMiddleware.AspNetCore.FabricContainer..ctor(ServiceCollection services) in C:\dev\ServiceFabricGateway\src\S-Innovations.ServiceFabric.RegistrationMiddleware.AspNetCore\FabricContainer.cs:line 172
   at SInnovations.ServiceFabric.GatewayService.Program.Main(String[] args) in C:\dev\ServiceFabricGateway\apps\S-Innovations.ServiceFabric.GatewayService\Program.cs:line 78adminuser@nt1vm000001:/mnt/sfroot/_App/S-Innovations.ServiceFabric.GatewayApplicationType_App6/S-InnoPkg.Code.1.0.0$ abric.GatewayServiceP
```

**Comment by ashishnegi** [comment]
@pksorensen Did you get this working ?

**Comment by yi-peng** [comment]
I met same issue. Any comments so far?

**Comment by ashishnegi** [comment]
@yi-peng Can you print your `env` in your entrypoint script and see what is the value of `LD_LIBRARY_PATH`. Why `EnvironmentVariablesConfigurationProvider` is getting multiple values of `LD_LIBRARY_PATH` ?

**Comment by yi-peng** [comment]
@ashishnegi I added a script to print env in a file during startup. I didn't see public IP for ubuntu nodes, how can I download that file from nodes?

**Comment by pksorensen** [comment]
I am in the main program setting the following code:

`Environment.SetEnvironmentVariable("LD_LIBRARY_PATH", null);`

I did that to get around it crashing.

**Comment by christopher-lawrence** [comment]
I'm having this exact issue, but I'm using dotnet core 2.2.

Setup: Use the base docker image provided in the OSX setup guide.  I added the installation of dotnet 2.2 runtime to the Docker file.  Create a base stateless service (created via the Yeoman templates).  Add a Kestrel listener.  Without the Kestrel listener, the service starts fine.

@ashishnegi I added a setup script to grab my env variables (below).  LD_LIBRARY_PATH is already set.

If I try to set the environment variable to null as @pksorensen suggested, I get the following error:

```
No service for type 'Microsoft.AspNetCore.Hosting.Server.IServer' has been registered.
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.EnsureServer()
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.BuildApplication()
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.StartAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.Start()
   at Microsoft.ServiceFabric.Services.Communication.AspNetCore.AspNetCoreCommunicationListener.OpenAsync(CancellationToken cancellationToken)
   at Microsoft.ServiceFabric.Services.Runtime.StatelessServiceInstanceAdapter.OpenCommunicationListenersAsync(CancellationToken cancellationToken)
   at Microsoft.ServiceFabric.Services.Runtime.StatelessServiceInstanceAdapter.System.Fabric.IStatelessServiceInstance.OpenAsync(IStatelessServicePartition partition, CancellationToken cancellationToken)

```

```
Fabric_ApplicationName=fabric:/demoApp
Fabric_CodePackageName=Code
HOSTNAME=329afcaae7e1
Fabric_ApplicationHostType=Activated_SingleCodePackage
FabricPackageFileName=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/Fabric.Package.current.xml
HostedServiceName=HostedService/N0010_Fabric
Fabric_Endpoint_ServiceEndpoint=80
Fabric_ServicePackageInstanceSeqNum=131902560055782122
Fabric_NodeIPOrFQDN=172.17.0.2
Fabric_Folder_App_Work=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App2/work
LC_ALL=en_US.UTF-8
Fabric_NodeName=N0010
FabricActivatorAddress=localhost:39011
Fabric_NodeId=c2e9eff19761acc9924422c53c8943d0
TEMP=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App2/temp
LD_LIBRARY_PATH=/home/ClusterDeployer/../FabricDrop/bin/Fabric/Fabric.Code
Fabric_Folder_App_Temp=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App2/temp
Fabric_Folder_App_Log=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App2/log
Fabric_Endpoint_IPOrFQDN_ServiceEndpoint=172.17.0.2
PATH=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/Fabric.Code:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
_=/usr/bin/printenv
Fabric_ApplicationId=demoAppType_App2
PWD=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App2/work
LANG=en_US.UTF-8
Fabric_IsCodePackageActivatorHost=false
Fabric_Folder_Application_OnHost=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App2
Fabric_ServicePackageName=DemoStatelessPkg
SHLVL=3
HOME=/root
LANGUAGE=en_US:en
Fabric_ServicePackageVersionInstance=1.0:1.0:131902636532217770
Fabric_RuntimeConnectionAddress=localhost:19006
TMP=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App2/temp
Fabric_IsContainerHost=false
Fabric_ApplicationHostId=c61b5323-f28e-7644-8e29-4bdf719afb05
Fabric_Folder_Application=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App2
Fabric_CodePackageInstanceSeqNum=131902560055782123
Fabric_ServicePackageActivationId=
```

**Comment by ashishnegi** [comment]
@christopher-lawrence 
There is no duplicate `LD_LIBRARY_PATH` in the `env` output you shared.
Setting `LD_LIBRARY_PATH` to null is not right. We set `LD_LIBRARY_PATH` to add our native dlls to the search path. This works fine with aspnet.core 2.0. 

Can you use reflection to enumerate `EnvironmentAugments.EnumerateEnvironmentVariables()` which has actual list of what `dotnet` is looking at ? Reference [here](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Extensions/src/System/Environment.cs#L34)
This will tell us exactly what dotnet greater than 2.0 are seeing.

**Comment by christopher-lawrence** [comment]
@ashishnegi 
As stated in a previous comment, calling `Environment.GetEnvironmentVariables()` causes the original exception.

I'm trying to figure out how to use reflection to get the environment variables.  I found the implementation to the `EnvironmentAugments.EnumerateEnvironmentVariables()` [here](https://github.com/dotnet/corert/blob/5beb4f805b13cce09eeafea9d2a9bad0fc0f1cb8/src/System.Private.CoreLib/src/Internal/Runtime/Augments/EnvironmentAugments.Unix.cs#L28), but I'm not sure how to gain access to the `Interop` class it is referencing.

**Comment by ashishnegi** [comment]
@christopher-lawrence Yes, it was clear to me that `environment.getenvironmentvariables()` causes original exception. However, as in your output of `env`, there is no duplicate `LD_LIBRARY_PATH`. So, why are we seeing that exception ?
Please use this code to print the env variable seen by dotnet :
```cs
            try
            {
                Environment.GetEnvironmentVariables();
            }
            catch (Exception ex) // eat the exception. Called so that dotnet load the assembly for us.
            {
                Console.WriteLine(ex);
            }

            var assemblyDll = Assembly.Load("System.Private.CoreLib.dll");
            var envArgType = assemblyDll.GetType("Internal.Runtime.Augments.EnvironmentAugments");
            var enumerateEnvVarMethod = envArgType.GetMethod("EnumerateEnvironmentVariables", new Type[] { });
            var envKVs = (IEnumerable<KeyValuePair<string, string>>) enumerateEnvVarMethod.Invoke(envArgType, null);
            foreach (var kv in envKVs)
            {
                Console.WriteLine($"{kv.Key} => {kv.Value}");
            }

            /// Outputs:
            /// ALLUSERSPROFILE => C:\ProgramData
            // CommonProgramFiles => C:\Program Files\Common Files
            // CommonProgramFiles(x86) => C:\Program Files(x86)\Common Files
            // CommonProgramW6432 => C:\Program Files\Common Files
            // ComSpec => C:\WINDOWS\system32\cmd.exe
            // DriverData => C:\Windows\System32\Drivers\DriverData
            // ENABLE_XAML_DIAGNOSTICS_SOURCE_INFO => 1
            // FPS_BROWSER_APP_PROFILE_STRING => Internet Explorer
            // FPS_BROWSER_USER_PROFILE_STRING => Default
            // HOMEDRIVE => C:
```

**Comment by christopher-lawrence** [comment]
@abhishekram that worked.  I wasn't sure which library to load to get to that class.  Thanks for the help!

Looks like there _is_ more than one LD_LIBRARY_PATH defined.

I'm wondering if [this](https://github.com/Microsoft/service-fabric/blob/c326b801c6c709f36684700edfe7bb88ceec9d7f/src/prod/src/Hosting2/ProcessActivator.Linux.cpp#L499) code is the culprit?  My C++ isn't great, I just searched the service fabric repo for the LD_LIBRARY_PATH variable.

```
[FabricActivatorAddress, localhost:46111]=localhost:46111
[FabricPackageFileName, /home/ClusterDeployer/ClusterData/Data/N0010/Fabric/Fabric.Package.current.xml]=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/Fabric.Package.current.xml
[Fabric_ApplicationHostId, 7e81df8b-36ab-204c-927d-64979131d985]=7e81df8b-36ab-204c-927d-64979131d985
[Fabric_ApplicationHostType, Activated_SingleCodePackage]=Activated_SingleCodePackage
[Fabric_ApplicationId, demoAppType_App1]=demoAppType_App1
[Fabric_ApplicationName, fabric:/demoApp]=fabric:/demoApp
[Fabric_CodePackageInstanceSeqNum, 131903979794222990]=131903979794222990
[Fabric_CodePackageName, Code]=Code
[Fabric_Endpoint_IPOrFQDN_ServiceEndpoint, 172.17.0.2]=172.17.0.2
[Fabric_Endpoint_ServiceEndpoint, 80]=80
[Fabric_Folder_Application, /home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1]=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1
[Fabric_Folder_Application_OnHost, /home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1]=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1
[Fabric_Folder_App_Log, /home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1/log]=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1/log
[Fabric_Folder_App_Temp, /home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1/temp]=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1/temp
[Fabric_Folder_App_Work, /home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1/work]=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1/work
[Fabric_IsCodePackageActivatorHost, false]=false
[Fabric_IsContainerHost, false]=false
[Fabric_NodeId, c2e9eff19761acc9924422c53c8943d0]=c2e9eff19761acc9924422c53c8943d0
[Fabric_NodeIPOrFQDN, 172.17.0.2]=172.17.0.2
[Fabric_NodeName, N0010]=N0010
[Fabric_RuntimeConnectionAddress, localhost:19006]=localhost:19006
[Fabric_ServicePackageActivationId, ]=
[Fabric_ServicePackageInstanceSeqNum, 131903979794222989]=131903979794222989
[Fabric_ServicePackageName, DemoStatelessPkg]=DemoStatelessPkg
[Fabric_ServicePackageVersionInstance, 1.0:1.0:131903984894530170]=1.0:1.0:131903984894530170
[HOME, /root]=/root
[HostedServiceName, HostedService/N0010_Fabric]=HostedService/N0010_Fabric
[HOSTNAME, f95ce12f5502]=f95ce12f5502
[LANG, en_US.UTF-8]=en_US.UTF-8
[LANGUAGE, en_US:en]=en_US:en
[LC_ALL, en_US.UTF-8]=en_US.UTF-8
[LD_LIBRARY_PATH, /home/ClusterDeployer/../FabricDrop/bin/Fabric/Fabric.Code]=/home/ClusterDeployer/../FabricDrop/bin/Fabric/Fabric.Code
[OLDPWD, /home/FabricDrop/bin/Fabric/Fabric.Code]=/home/FabricDrop/bin/Fabric/Fabric.Code
[PATH, /home/ClusterDeployer/ClusterData/Data/N0010/Fabric/Fabric.Code:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/Fabric.Code:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
[PWD, /home/FabricDrop/bin]=/home/FabricDrop/bin
[SHLVL, 2]=2
[TEMP, /home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1/temp]=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1/temp
[TMP, /home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1/temp]=/home/ClusterDeployer/ClusterData/Data/N0010/Fabric/work/Applications/demoAppType_App1/temp
[_, ./FabricHost]=./FabricHost
[LD_LIBRARY_PATH, /home/ClusterDeployer/../FabricDrop/bin/Fabric/Fabric.Code]=/home/ClusterDeployer/../FabricDrop/bin/Fabric/Fabric.Code
```

**Comment by ashishnegi** [comment]
@christopher-lawrence thanks for output.
@chenxumsft Please see that we are adding `LD_LIBRARY_PATH` twice in env variables for the process and it is crashing .net core 2.1 linux apps.

**Comment by ashishnegi** [comment]
@christopher-lawrence / @yi-peng / @pksorensen / @git-heath 
We have created an internal bug to fix adding multiple `LD_LIBRARY_PATH`.
However, many linux tools handle this gracefully and so does dotnetn 2.0 also handled it gracefully.
This started failing in dotnet 2.1 / 2.2.

As per this [corefx/issues/34263#issuecomment-450406251](https://github.com/dotnet/corefx/issues/34263#issuecomment-450406251) , can you switch to dotnet 3.0 preview-1 ? If there are any problems switching to 3.0, please raise the concerns in corefx thread.

**Comment by christopher-lawrence** [comment]
@ashishnegi At the start of the service (either in Program.cs or Service.cs), if you do call `Environment.SetEnvironmentVariable("LD_LIBRARY_PATH", null);` this will allow the service to start up.  I had an error in my code in previously when I tried this.

I confirmed this will remove the last LD_LIBRARY_PATH that is set.  This is a workaround that I'm fine with until 3.0 is official.

**Comment by pksorensen** [comment]
Thanks for confirming @christopher-lawrence - i am also using this work around until 3.0

**Comment by luisgoncalves** [comment]
@ashishnegi 
> We have created an internal bug to fix adding multiple `LD_LIBRARY_PATH`.

Do you have any idea when this will be patched? I'm assuming this is something on the SF "server runtime".. We have a bunch of SF apps and it's not very practical to add the mentioned workaround on all the services.

Thanks.

**Comment by shsha-msft** [comment]
Thanks for reporting this issue. We have fixed it and it will be part of the upcoming 6.4 CU5 (fifth refresh) release.
